---
title: "Homework 3"
author: "Yavuz Delen"
date: "4/14/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../')) 
```


#Answer2
```{r}
X <- matrix(c(5,6, 3,-4), nrow = 2, byrow = T)
Y <- matrix(c(-12/19, 39/38), nrow = 2, byrow = T)
X %*% Y
```




#Answer3
```{r}
pheno_data <- read.csv("data/DATA_rootnum_age10_rootnum.csv", header = TRUE, stringsAsFactors = FALSE)
#View(pheno_data)
#hist(pheno_data$Derregressed_BV)
#hist(pheno_data$Derregressed_BV, xlab="pheno_data$Derregressed_BV", col="red", breaks=100)
geno_data <- "https://jyanglab.com/img/data/Snp_Data.csv"
loblolly_geno <- read.csv(geno_data, header=TRUE, stringsAsFactors = FALSE)
#head(loblolly_geno)
#View(loblolly_geno)
dim(pheno_data) #926 4
dim(loblolly_geno) #926 4854
loblolly_geno[1:5, 1:5]
```


## Remove missing phenotypes
```{r, eval=FALSE}
is.na(pheno_data$Derregressed_BV)
#sum(is.na(pheno_data$Derregressed_BV)) #1
na.index <-  which(is.na(pheno_data$Derregressed_BV))
#head(pheno_data[na.index,])
length(na.index) #1
pheno_data <- pheno_data[-na.index, ]
#select genotypic dataframe by using phenotypic data
loblolly_geno <- loblolly_geno[loblolly_geno$Genotype %in% pheno_data$Genotype, ]
# phenotypes 
y <- pheno_data$Derregressed_BV
y <- matrix(y, ncol=1)
y
# markers 
loblolly_geno <- loblolly_geno[,-1]  
loblolly_geno[loblolly_geno == -9] <- NA
dim(loblolly_geno) #925 4853
```


# SNP quality control
### Missingness and MAF

```{r, eval=FALSE, echo=TRUE}
# missing rate
missing_snp <- apply(loblolly_geno, 2, function(x){sum(is.na(x))/length(x)})
head(missing_snp)
#length(missing_snp) #4853
#hist(missing_snp, breaks=25)
# minor allele frequency
maf <- apply(loblolly_geno, 2, function(x){
  frq <- mean(x, na.rm=TRUE)/2
  return(ifelse(frq > 0.5, 1-frq, frq))
})
```

#### Plot the results
```{r, eval=FALSE, echo=TRUE}
hist(missing_snp, breaks=50, col="red", xlab="SNP Missing rate")
hist(maf, breaks=50, col="red", xlab="Minor Allele Freq")
```


# SNP quality control

Removing SNPs with high missing rate (missingness > 0.2) and low MAF (MAF < 0.05)

```{r, eval=FALSE, echo=TRUE}
idx1 <- which(missing_snp > 0.2) #155
length(idx1)
idx2 <- which(maf < 0.05) #1640
length(idx2)
idx <- unique(c(idx1, idx2)) #1779
length(idx)
loblolly_geno2 <- loblolly_geno[, -idx]
dim(loblolly_geno2)  #925 3074
loblolly_geno2[1:10, 1:5]
```


### Missing marker imputation

Replace missing marker genotypes with __mean values__. Then store the marker genotypes in a matrix object `Z`. 

```{r, eval=FALSE, echo=TRUE}
Z <- matrix(0, ncol=ncol(loblolly_geno2), nrow=nrow(loblolly_geno2))
for (j in 1:ncol(loblolly_geno2)){
  #cat("j = ", j, '\n')
  Z[,j] <- ifelse(is.na(loblolly_geno2[,j]), mean(loblolly_geno2[,j], na.rm=TRUE), loblolly_geno2[,j])
}
Z[1:10, 1:5]
sum(is.na(Z))
```


# Genomic relationship

### SNP Matrix standardization

Standardize the genotype matrix to have a mean of zero and variance of one. Save this matrix as `Zs`. 

```{r, eval=FALSE, echo=TRUE}
Zs <- scale(Z, center = TRUE, scale = TRUE)
#Zs[1:10, 1:5]
n <- nrow(Zs)
n
m <- ncol(Zs)
m
```


### Calcualte genomic relationship
 

```{r, eval=FALSE, echo=TRUE}
# Given matrices x and y as arguments, return a matrix cross-product. This is formally equivalent to (but usually slightly faster than) the call t(x) %*% y (crossprod) or x %*% t(y) (tcrossprod).
G <- tcrossprod(Zs) / ncol(Zs)
#G <- Zs %*% t(Zs) / ncol(Zs)
G <- G + diag(n)*0.001
```



# Solve MME for GBLUP

```{r, eval=FALSE, echo=TRUE}
lambda <- 4.087116 # fit$Ve / fit$Vu
Ginv <- solve(G)
ones <- matrix(1, ncol=1, nrow=n)
Z <- diag(n)
# Given matrices x and y as arguments, return a matrix cross-product. This is formally equivalent to (but usually slightly faster than) the call t(x) %*% y (crossprod) or x %*% t(y) (tcrossprod).
LHS1 <- cbind(crossprod(ones), crossprod(ones, Z)) 
LHS2 <- cbind(crossprod(Z, ones), crossprod(Z) +  Ginv*lambda)
LHS <- rbind(LHS1, LHS2)
RHS <- rbind( crossprod(ones, y), crossprod(Z,y) )
sol <- solve(LHS, RHS)
head(sol)
tail(sol)
```



```{r, eval=FALSE, echo=TRUE}
#install.packages("rrBLUP")
library(rrBLUP)
fit <- mixed.solve(y = y, K=G)
# additive genetic variance
fit$Vu
# residual variance
fit$Ve
# intercept 
fit$beta
# additive genetic values
head(fit$u)
tail(fit$u)
# genomic h2
fit$Vu / (fit$Vu + fit$Ve)
# ratio of variance components 
fit$Ve / fit$Vu
plot(x=sol[-1], y=fit$u)
```


## RR-BLUP
```{r, eval=FALSE, echo=TRUE}
lambda <- 12566.89 # fit$Ve / fit$Vu
ones <- matrix(1, ncol=1, nrow=n)
I <- diag(m)
LHS1 <- cbind(crossprod(ones), crossprod(ones, Zs)) 
LHS2 <- cbind(crossprod(Zs, ones), crossprod(Zs) +  I*lambda)
LHS <- rbind(LHS1, LHS2)
RHS <- rbind( crossprod(ones, y), crossprod(Zs,y) )
sol2 <- solve(LHS, RHS)
head(sol2)
tail(sol2)
#dim(sol2) #3075
```


```{r, eval=FALSE, echo=TRUE}
library(rrBLUP)
fit2 <- mixed.solve(y = y, Z=Zs)
# marker additive genetic variance
fit2$Vu
# residual variance
fit2$Ve
# intercept 
fit2$beta
# marker additive genetic effects
head(fit2$u)
tail(fit2$u)
# ratio of variance components 
fit2$Ve / fit2$Vu
plot(x=sol2[-1], y=fit2$u)
cor(x=sol2[-1], y=fit2$u)
```


## 10-fold cross validation
```{r, eval=FALSE, echo=TRUE}
n.trn <- 833
n.tst <- 92
y.trn <- y[1:n.trn]
y.tst <- y[n.trn+1:n.tst]
Zs.trn <- Zs[1:n.trn,]
Zs.tst <- Zs[n.trn+1:n.tst,]
Gtrn <- tcrossprod(Zs.trn) / ncol(Zs.trn)
Gtrn <- Gtrn + diag(n.trn)*0.001
Gtst.trn <- tcrossprod(Zs.tst, Zs.trn) / ncol(Zs.tst)
#Gtrn <- G[1:n.trn, 1:n.trn]
#Gtst.trn <- G[n.trn+1:n.tst, 1:n.trn]
lambda <- 12566.89 # fit$Ve / fit$Vu
Ginv.trn <- solve(Gtrn)
ones <- matrix(1, ncol=1, nrow=n.trn)
Z <- diag(n.trn)
LHS1 <- cbind(crossprod(ones), crossprod(ones, Z)) 
LHS2 <- cbind(crossprod(Z, ones), crossprod(Z) +  Ginv.trn*lambda)
LHS <- rbind(LHS1, LHS2)
RHS <- rbind( crossprod(ones, y.trn), crossprod(Z,y.trn) )
sol.trn <- solve(LHS, RHS)
# prediction
y.hat <- Gtst.trn %*% Ginv.trn %*% matrix(sol.trn[c(2:(n.trn+1))])
cor(y.hat, y[(n.trn+1):n])
plot(y.hat, y[(n.trn+1):n])
```





```{r, eval=FALSE, echo=TRUE}
Zs.trn <- Zs[1:n.trn, ]
Zs.tst <- Zs[n.trn+1:n.tst, ]
lambda <- 4326.212 # fit$Ve / fit$Vu
ones <- matrix(1, ncol=1, nrow=n.trn)
I <- diag(m)
LHS1 <- cbind(crossprod(ones), crossprod(ones, Zs.trn)) 
LHS2 <- cbind(crossprod(Zs.trn, ones), crossprod(Zs.trn) +  I*lambda)
LHS <- rbind(LHS1, LHS2)
RHS <- rbind( crossprod(ones, y.trn), crossprod(Zs.trn, y.trn) )
sol.trn <- solve(LHS, RHS)
# prediction
y.hat2 <- Zs.tst %*% matrix(sol.trn[-1])
head(y.hat2)
dim(y.hat2)
# cor(y.hat2, y[(n.trn+1):n])
plot(y.hat2, y[(n.trn+1):n])
```




#Answer4

```{r}
library("ggplot2")
boxplot(y.hat, xlab = "GBLUP", ylab = "Prediction", main = "Prediction Accuracy")
boxplot(y.hat2, xlab = "rrBLUP", ylab = "Prediction", main = "Prediction Accuracy")
comb <- cbind(y.hat, y.hat2)
head(comb)
boxplot(comb, xlab = c("1.GBLUP", "2.rrBLUP"), ylab = "Prediction", main = "Prediction Accuracy Results")


#library("ggplot2")
#prediction <- rbind(y.hat, y.hat2)
#ggplot(prediction, aes(x=feature, y=Pairwise, fill=feature)) + 
#  geom_violin(trim=FALSE)+
#  labs(title="Prediction Accuracy Results", x="", y = "Prediction")+
#  geom_boxplot(width=0.1, fill="white")+
#  scale_fill_brewer(palette="Blues") + 
#  theme_classic()
```